### PR TITLE
Add SRNE ASF config example

### PR DESCRIPTION
**Pull request by syssi**
See #28

### CODE REVIEW

**Comment by felipecrs** [comment]
@syssi, my RS485-TTL converter finally arrived.

I did the wiring, but I was not able to read any data still. This is the diagram I used:

image

This is how it ended up like:



This is my config:

[monitor-do-inversor.yaml.txt](https://github.com/user-attachments/files/17787353/monitor-do-inversor.yaml.txt)

And this is what I found in the logs:

```
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=256 count=3
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=263 count=3
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=270 count=4
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=516 count=1
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=531 count=3
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61496 count=1
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61498 count=1
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61501 count=1
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61503 count=1
[14:26:21][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61512 count=1
[14:26:21][D][uart_debug:114]: >>> 01:03:F0:48:00:01:37:1C
[14:26:21][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:21][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:21][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:21][D][uart_debug:114]: >>> 01:03:F0:48:00:01:37:1C
[14:26:21][D][modbus_controller:038]: Modbus command to device=1 register=0xF048 no response received - removed from send queue
[14:26:21][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][W][daly_bms_ble:101]: [28:BB:ED:00:08:24] Not connected
[14:26:22][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:22][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][modbus_controller:038]: Modbus command to device=1 register=0x100 no response received - removed from send queue
[14:26:22][D][uart_debug:114]: >>> 01:03:01:07:00:03:B5:F6
[14:26:23][D][uart_debug:114]: >>> 01:03:01:07:00:03:B5:F6
[14:26:23][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:23][D][uart_debug:114]: >>> 01:03:01:07:00:03:B5:F6
[14:26:23][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:23][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:23][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:23][D][uart_debug:114]: >>> 01:03:01:07:00:03:B5:F6
[14:26:23][D][uart_debug:114]: >>> 01:03:01:07:00:03:B5:F6
[14:26:23][D][modbus_controller:038]: Modbus command to device=1 register=0x107 no response received - removed from send queue
[14:26:23][D][uart_debug:114]: >>> 01:03:01:0E:00:04:24:36
[14:26:24][D][uart_debug:114]: >>> 01:03:01:0E:00:04:24:36
[14:26:24][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:24][D][uart_debug:114]: >>> 01:03:01:0E:00:04:24:36
[14:26:24][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:24][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:24][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:24][D][uart_debug:114]: >>> 01:03:01:0E:00:04:24:36
[14:26:24][D][uart_debug:114]: >>> 01:03:01:0E:00:04:24:36
[14:26:24][D][modbus_controller:038]: Modbus command to device=1 register=0x10E no response received - removed from send queue
[14:26:25][D][uart_debug:114]: >>> 01:03:02:04:00:01:C4:73
[14:26:25][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:25][D][uart_debug:114]: >>> 01:03:02:04:00:01:C4:73
[14:26:25][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:25][D][uart_debug:114]: >>> 01:03:02:04:00:01:C4:73
[14:26:25][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:25][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:25][D][uart_debug:114]: >>> 01:03:02:04:00:01:C4:73
[14:26:25][D][uart_debug:114]: >>> 01:03:02:04:00:01:C4:73
[14:26:25][D][modbus_controller:038]: Modbus command to device=1 register=0x204 no response received - removed from send queue
[14:26:26][D][uart_debug:114]: >>> 01:03:02:13:00:03:F5:B6
[14:26:26][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:26][D][uart_debug:114]: >>> 01:03:02:13:00:03:F5:B6
[14:26:26][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:26][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:26][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:26][D][uart_debug:114]: >>> 01:03:02:13:00:03:F5:B6
[14:26:26][D][uart_debug:114]: >>> 01:03:02:13:00:03:F5:B6
[14:26:26][D][uart_debug:114]: >>> 01:03:02:13:00:03:F5:B6
[14:26:27][D][modbus_controller:038]: Modbus command to device=1 register=0x213 no response received - removed from send queue
[14:26:27][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:27][D][uart_debug:114]: >>> 01:03:F0:38:00:01:36:C7
[14:26:27][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:27][D][uart_debug:114]: >>> 01:03:F0:38:00:01:36:C7
[14:26:27][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:27][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:27][D][uart_debug:114]: >>> 01:03:F0:38:00:01:36:C7
[14:26:27][D][uart_debug:114]: >>> 01:03:F0:38:00:01:36:C7
[14:26:28][D][uart_debug:114]: >>> 01:03:F0:38:00:01:36:C7
[14:26:28][D][modbus_controller:038]: Modbus command to device=1 register=0xF038 no response received - removed from send queue
[14:26:28][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:28][D][uart_debug:114]: >>> 01:03:F0:3A:00:01:97:07
[14:26:28][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:28][D][uart_debug:114]: >>> 01:03:F0:3A:00:01:97:07
[14:26:28][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:28][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:28][D][uart_debug:114]: >>> 01:03:F0:3A:00:01:97:07
[14:26:28][D][uart_debug:114]: >>> 01:03:F0:3A:00:01:97:07
[14:26:29][D][uart_debug:114]: >>> 01:03:F0:3A:00:01:97:07
[14:26:29][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:29][D][modbus_controller:038]: Modbus command to device=1 register=0xF03A no response received - removed from send queue
[14:26:29][D][uart_debug:114]: >>> 01:03:F0:3D:00:01:26:C6
[14:26:29][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:29][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:29][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:29][D][uart_debug:114]: >>> 01:03:F0:3D:00:01:26:C6
[14:26:29][D][uart_debug:114]: >>> 01:03:F0:3D:00:01:26:C6
[14:26:29][D][uart_debug:114]: >>> 01:03:F0:3D:00:01:26:C6
[14:26:30][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
[14:26:30][D][uart_debug:114]: >>> 01:03:F0:3D:00:01:26:C6
[14:26:30][D][modbus_controller:038]: Modbus command to device=1 register=0xF03D no response received - removed from send queue
[14:26:30][D][sensor:094]: 'Grid Power': Sending state nan W with 0 decimals of accuracy
[14:26:30][D][uart_debug:114]: >>> 01:03:F0:3F:00:01:87:06
[14:26:30][D][sensor:094]: 'Solar Power': Sending state nan W with 0 decimals of accuracy
[14:26:30][D][sensor:094]: 'Inverter Load': Sending state nan W with 0 decimals of accuracy
[14:26:30][D][uart_debug:114]: >>> 01:03:F0:3F:00:01:87:06
[14:26:30][D][uart_debug:114]: >>> 01:03:F0:3F:00:01:87:06
[14:26:31][D][uart_debug:114]: >>> 01:03:F0:3F:00:01:87:06
[14:26:31][D][sensor:094]: 'Battery Power': Sending state nan W with 1 decimals of accuracy
```

Do you have any idea?

I'm sure the registers are correct, because I am able to read the inverter data through USB through [PythonProtocolGateway](https://github.com/HotNoob/PythonProtocolGateway/blob/main/protocols/srne/srne_v1.7.holding_registry_map.csv), which uses the same registers.

**Comment by syssi** [comment]
`` 
[14:26:21][D][uart_debug:114]: >>> 01:03:F0:48:00:01:37:1C
[14:26:21][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
[14:26:22][D][uart_debug:114]: >>> 01:03:01:00:00:03:04:37
```

The `>>>` indicates outgoing traffic/requests. We don't see any respone / incoming traffic (`<<<`). 

At your YAML configuration you are using GPIO43 & GPIO44 as TX/RX pins:

```
uart:
  - id: uart_0
    tx_pin: GPIO43
    rx_pin: GPIO44
    baud_rate: 9600
```

At your photo the RS485 converter is attached to RXD (GPIO3) and TXD (GPIO1). Could you attached the converter to GPIO43/GPIO44 and provide another log?

**Comment by felipecrs** [comment]
In fact these pins number are correct for my board:

https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html

I was also able to connect it to the Daly BMS UART directly (through the ESPHome Daly HKMS PR), and it worked fine.

So I really don't think that's the issue.

But the fact there's no incoming data, that's something to think about. I hope my inverter isn't malfunctioning.

**Comment by syssi** [comment]
Another idea: Do you have a second ESP and RS485 converter to make sure the setup works in general?

```
ESP1 <--> RS485 converter1 <-----> RS485 converter2 <--> ESP2
```

**Comment by felipecrs** [comment]
I have ESP8266 lying around, and a couple of RS485-TTL converters to spare. But I'm a bit unsure how to setup this test up.

Anyway, let me double check this before going that route. Thanks for suggesting this option.

The only guide I found about this connection is this one:

https://community.home-assistant.io/t/srne-inverter-integration-help/609730

And it has a few gotchas I don't understand:

1. Why removing R7 from the RS485-TTL converter
2. Why jumpering DE and RE
3. Why your config doesn't mention any flow control pin, while that one mentions

There's also a chance that guide doesn't apply to my inverter, even though it is theoretically to the same series as mine (their is SRNE HF4850S80-H, mine is SRNE HF2430U60-100).

**Comment by felipecrs** [comment]
@syssi, pretty much everything is working as expected. I just did some adjustments:

1. I changed multiply factor from 0.4 to 0.2 and 2.5 to 5, as my inverter is 24V.
   - `xE003` reads the System Voltage by the way. It would return 24V in my inverter, and 48V in others. This should be enough to make these maths automatically, although probably not worth it.
2. Adjusted the number entities min and max accordingly too, both voltages and current.

Another feedback:

3. Serial number isn't being read properly:

	image
	
	Note that PythonProtocolGateway do read these numbers ok, but I'm not sure it's the same registers (`0x0035~0x0048`).

File: esp32-srne-asf-example.yaml [hunk 1]
```diff
@@ -0,0 +1,822 @@
...
+select:
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "Output Priority"
+    id: output_priority
+    address: 0xE204
+    value_type: U_WORD
+    optionsmap:
+      "Solar": 0
+      "Utlity": 1
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:712]:
```suggestion
      "Utility": 1
```

File: esp32-srne-asf-example.yaml [hunk 2]
```diff
@@ -0,0 +1,822 @@
...
+    min_value: 40
+    max_value: 60
+    step: 0.1
+    lambda: "return x * 0.4;"
+    write_lambda: "return x * 2.5;"
+
+  - platform: modbus_controller
+    modbus_controller_id: smg0
+    name: "Bulk Charge Voltage"
+    id: bulk_voltage
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:652]:
```suggestion
    name: "Boost Charge Voltage"
    id: boost_voltage
```

File: esp32-srne-asf-example.yaml [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+    modbus_controller_id: smg0
+    name: "Grid Charge Current"
+    address: 0xE205
+    value_type: U_WORD
+    register_type: holding
+    unit_of_measurement: "A"
+    min_value: 0
+    max_value: 120
+    step: 1
+    multiply: 0.1
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:688]:
```suggestion
    lambda: "return x * 0.1;"
    write_lambda: "return x * 10;"
```

File: esp32-srne-asf-example.yaml [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+    name: "Charge Priority"
+    address: 0xE20F
+    value_type: U_WORD
+    optionsmap:
+      "Solar and Utility": 0
+      "Utility-Solar": 1
+      "Solar-Utility": 2
+      "Solar Only": 3
+
+  - platform: modbus_controller
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:763]: I suggest converting this one to switch entity.

File: esp32-srne-asf-example.yaml [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+    modbus_controller_id: smg0
+    name: "ECO Mode"
+    id: ecomode
+    address: 0xE20C
+    value_type: U_WORD
+    optionsmap:
+      "Disable": 0
+      "Enable": 1
+
+  - platform: modbus_controller
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:734]: I suggest converting this one to switch entity.

File: esp32-srne-asf-example.yaml [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+    modbus_controller_id: smg0
+    name: "AC Input Voltage Range"
+    address: 0xE20B
+    value_type: U_WORD
+    optionsmap:
+      "Appliance": 0
+      "UPS": 1
+
+  - platform: modbus_controller
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:724]: I suggest converting this one to switch entity.

**Comment by felipecrs** [comment]
Something I noticed in my ESPHome logs is a bunch of `CRC Check` errors. Example:

```
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 7E4E!=6A7E
[01:56:47][W][modbus_controller:027]: Modbus device=1 set offline
[01:56:47][D][modbus_controller:038]: Modbus command to device=1 register=0xF03A no response received - removed from send queue
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=301
[01:56:47][D][sensor:094]: 'Inverter Grid Power': Sending state 496.44006 W with 0 decimals of accuracy
[01:56:47][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:47][W][modbus:124]: Modbus CRC Check failed! C0C1!=203
[01:56:48][D][sensor:094]: 'Inverter Battery Power': Sending state -0.00000 W with 1 decimals of accuracy
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 92C9!=86F9
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 804D!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! C0C1!=203
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=1FF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! A080!=F907
[01:56:48][D][sensor:094]: 'Inverter Grid Power': Sending state 496.44006 W with 0 decimals of accuracy
[01:56:48][D][modbus_controller:038]: Modbus command to device=1 register=0xF03D no response received - removed from send queue
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:48][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 5088!=44B8
[01:56:49][D][sensor:094]: 'Inverter Battery Power': Sending state -0.00000 W with 1 decimals of accuracy
[01:56:49][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61503 count=1
[01:56:49][W][modbus_controller:182]: Duplicate modbus command found: type=0x3 address=61512 count=1
[01:56:49][D][sensor:094]: 'WiFi signal dB': Sending state -11.00000 dBm with 0 decimals of accuracy
[01:56:49][D][sensor:094]: 'WiFi signal percent': Sending state 100.00000 % with 0 decimals of accuracy
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:49][W][modbus:124]: Modbus CRC Check failed! C0C1!=203
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=1FF
[01:56:49][W][modbus:124]: Modbus CRC Check failed! A080!=B800
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:49][D][sensor:094]: 'Inverter Grid Power': Sending state 496.44006 W with 0 decimals of accuracy
[01:56:49][W][modbus:124]: Modbus CRC Check failed! 4040!=301
[01:56:49][W][modbus:124]: Modbus CRC Check failed! D0!=44B8
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:50][W][modbus_controller:063]: Modbus device=1 back online
[01:56:50][D][modbus_controller.sensor:025]: Sensor new state: 0.00
[01:56:50][D][sensor:094]: 'Inverter Bypass Time': Sending state 0.00000 min with 1 decimals of accuracy
[01:56:50][D][sensor:094]: 'Inverter Battery Power': Sending state -0.00000 W with 1 decimals of accuracy
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=1FF
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:50][W][modbus:124]: Modbus CRC Check failed! A94F!=BD7F
[01:56:50][D][sensor:094]: 'Inverter Grid Power': Sending state 496.44006 W with 0 decimals of accuracy
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:50][W][modbus:124]: Modbus CRC Check failed! 4040!=1FF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][D][sensor:094]: 'Inverter Battery Power': Sending state -0.00000 W with 1 decimals of accuracy
[01:56:51][D][modbus_controller.sensor:025]: Sensor new state: 2461.00
[01:56:51][D][sensor:094]: 'Inverter Accumulated Grid Power': Sending state 246.10001 kWh with 1 decimals of accuracy
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 6808!=D01
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][D][modbus_controller.sensor:025]: Sensor new state: 93.00
[01:56:51][D][sensor:094]: 'Inverter Battery SOC': Sending state 93.00000 % with 0 decimals of accuracy
[01:56:51][D][modbus_controller.sensor:025]: Sensor new state: 269.00
[01:56:51][D][sensor:094]: 'Inverter Battery Voltage': Sending state 26.90000 V with 1 decimals of accuracy
[01:56:51][D][modbus_controller.sensor:025]: Sensor new state: -87.00
[01:56:51][D][sensor:094]: 'Inverter Battery Current': Sending state 8.70000 A with 1 decimals of accuracy
[01:56:51][D][daly_bms_ble:575]: Send command (handle 0x15): D2.03.00.00.00.3E.D7.B9 (8)
[01:56:51][D][esp32_ble_client:110]: [0] [28:BB:ED:00:08:24] ESP_GATTC_WRITE_CHAR_EVT
[01:56:51][D][sensor:094]: 'Inverter Grid Power': Sending state 496.44006 W with 0 decimals of accuracy
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][W][modbus:124]: Modbus CRC Check failed! 4040!=FFFF
[01:56:51][D][text_sensor:064]: 'Inverter Charge State': Sending state 'Quick Charge'
```

Do you have any idea if this is normal?

**Comment by felipecrs** [comment]
Last question today: is it normal that these green and blue leds keeps flashing so much? I wonder whether this won't wear them.


https://github.com/user-attachments/assets/f01d89c1-58b6-4d04-9ec7-53d1c44bc1f5

**Comment by syssi** [comment]
There is a LED attached to the TXD pin of your devboard. If you use/configure any other pin your could avoid the LED.

**Comment by syssi** [comment]
TODOs: Fix serial number. PythonProtocolGateway do read these numbers ok, but I'm not sure it's the same registers (0x0035~0x0048).

**Comment by felipecrs** [comment]
> TODOs: Fix serial number. PythonProtocolGateway do read these numbers ok, but I'm not sure it's the same registers (0x0035~0x0048).

@syssi, looks like you based yourself in a protocol which is different to my inverter's. Maybe you should keep this part as is. PythonProtocolGateway also attempts to use these registers when using SRNE protocol 3.9. Mine protocol is 1.96.

File: esp32-srne-asf-example.yaml [hunk 7]
```diff
@@ -0,0 +1,827 @@
...
+    id: grid_frequency
+    name: "Grid Frequency"
+    address: 0x215
+    unit_of_measurement: "Hz"
+    register_type: holding
+    value_type: U_WORD
+    accuracy_decimals: 1
+    filters:
+      - multiply: 0.01
+
```
Comment (felipecrs) [inline esp32-srne-asf-example.yaml:276]: I believe you are missing these two important registers, which are the inverter load voltage (phase A) and load current (phase A).

My inverter has only one phase, so only phase A reads data.

```suggestion

  - platform: modbus_controller
    modbus_controller_id: srne_0
    id: inverter_voltage
    name: "Inverter Voltage"
    address: 0x216
    unit_of_measurement: "V"
    register_type: holding
    value_type: U_WORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.1

  - platform: modbus_controller
    modbus_controller_id: srne_0
    id: inverter_current
    name: "Inverter Current"
    address: 0x219
    unit_of_measurement: "A"
    register_type: holding
    value_type: U_WORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.1
```

Even though [protocol 3.9](https://github.com/HotNoob/PythonProtocolGateway/blob/main/protocols/srne/srne_v3.9.holding_registry_map.csv) doesn't mention them, I suppose they should work, given the 0x218 (inverter frequency) works, which seems also not part of the 3.9 manual.

Comment (syssi) [reply esp32-srne-asf-example.yaml:276]: Added.

**Comment by felipecrs** [comment]
> Something I noticed in my ESPHome logs is a bunch of `CRC Check` errors.

In case someone else reach the same issue: https://community.home-assistant.io/t/srne-inverter-integration-help/609730/49?u=felipecrs

**Pull request merged**